src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -1269,10 +1269,36 @@ public int Send(ReadOnlySpan<byte> buffer, SocketFlags socketFlags, out SocketEr
 
         public void SendFile(string? fileName)
         {
-            SendFile(fileName, null, null, TransmitFileOptions.UseDefaultWorkerThread);
+            SendFile(fileName, ReadOnlySpan<byte>.Empty, ReadOnlySpan<byte>.Empty, TransmitFileOptions.UseDefaultWorkerThread);
         }
 
         public void SendFile(string? fileName, byte[]? preBuffer, byte[]? postBuffer, TransmitFileOptions flags)
+        {
+            SendFile(fileName, preBuffer.AsSpan(), postBuffer.AsSpan(), flags);
+        }
+
+        /// <summary>
+        /// Sends the file <paramref name="fileName"/> and buffers of data to a connected <see cref="Socket"/> object
+        /// using the specified <see cref="TransmitFileOptions"/> value.
+        /// </summary>
+        /// <param name="fileName">
+        /// A <see cref="string"/> that contains the path and name of the file to be sent. This parameter can be <see langword="null"/>.
+        /// </param>
+        /// <param name="preBuffer">
+        /// A <see cref="byte"/> array that contains data to be sent before the file is sent. This parameter can be <see langword="null"/>.
+        /// </param>
+        /// <param name="postBuffer">
+        /// A <see cref="byte"/> array that contains data to be sent after the file is sent. This parameter can be <see langword="null"/>.
+        /// </param>
+        /// <param name="flags">
+        /// One or more of <see cref="TransmitFileOptions"/> values.
+        /// </param>
+        /// <exception cref="ObjectDisposedException">The <see cref="Socket"/> object has been closed.</exception>
+        /// <exception cref="NotSupportedException">The <see cref="Socket"/> object is not connected to a remote host.</exception>
+        /// <exception cref="InvalidOperationException">The <see cref="Socket"/> object is not in blocking mode and cannot accept this synchronous call.</exception>
```

Comment:
Similarly, do we have test that we get InvalidOperationException if you try to do this with a non blocking socket?

Reply:
There's https://github.com/dotnet/runtime/blob/5a4483eb1c0918542298d1864b7233f962952880/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs#L213 

Trying something like
```c#
using var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
using var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
listener.BindToAnonymousPort(IPAddress.Loopback);
listener.Listen(1);

client.Connect(listener.LocalEndPoint);
using Socket server = listener.Accept();
server.ForceNonBlocking(force: true);

Assert.Throws<InvalidOperationException>(() => server.SendFile(null));
```
won't throw the exception (i.e. that test fails) on Windows and Linux.
After `ForceNonBlocking` both `_willBlock` and `_willBlockInternal` are `false`, so https://github.com/dotnet/runtime/blob/16e6e9530687b34cc93a85d4043093f06f9b7eff/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4045-L4051 won't throw.

Reply:
I don't mean ForceNonBlocking. That puts the socket into sync-over-async mode.

I mean literally a nonblocking socket, i.e. socket.Blocking = false.

Reply:
Tried this too, same result where the above `if` in ValidateBlockingMode won't throw.

![grafik](https://user-images.githubusercontent.com/5755834/105394502-f63e0a00-5c1d-11eb-9049-51b8569d703d.png)

Reply:
Huh. It seems like we don't actually prevent you from doing a sync SendFile in non-blocking mode.

Does it actually work? What happens if you send a really big file, does it block or return E_WOULDBLOCK or what?

I think we probably should be preventing this, but it looks like we are not currently...

Reply:
The following program sends 1000 MiB and runs just fine.
With the debugger it can be verifyed that it blocks until the send is finished (will step over once the client has read the sent data). I.e. it's the same behavior as when `server.Blocking = true`.

<details>
  <summary>SendFile non-blocking 1000 MiB</summary>

```c#
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading.Tasks;

using var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
using var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
listener.BindToAnonymousPort(IPAddress.Loopback);
listener.Listen(1);

client.Connect(listener.LocalEndPoint);
using Socket server = listener.Accept();
server.Blocking = false;

Task clientTask = RunClientAsync();

string fileName = Path.GetTempFileName();
try
{
    const long FileLength = 100 * 1024 * 1024;
    using (FileStream fs = File.Create(fileName))
    {
        fs.SetLength(FileLength);
    }

    Debug.Assert(!server.Blocking);

    server.SendFile(fileName);
    server.Close();
}
catch (SocketException ex)
{
    Console.WriteLine(ex.Message);
    Console.WriteLine(ex.SocketErrorCode);
}
finally
{
    File.Delete(fileName);
}

await clientTask;

Task RunClientAsync() => Task.Run(() =>
{
    byte[] buffer = new byte[4096];
    long totalReceived = 0;
    int received = 0;
    while ((received = client.Receive(buffer)) > 0)
    {
        Console.Write('.');
        totalReceived += received;
    }

    Console.WriteLine($"\nClient received {totalReceived / 1024 / 1024} MiB");
});

internal static class SocketTestExtensions
{
    // Binds to an IP address and OS-assigned port. Returns the chosen port.
    public static int BindToAnonymousPort(this Socket socket, IPAddress address)
    {
        socket.Bind(new IPEndPoint(address, 0));
        return ((IPEndPoint)socket.LocalEndPoint).Port;
    }
}
```
</details>

Reply:
Is that on Linux or Windows?

Reply:
Windows. Will try Linux soon

Reply:
On linux it throws:
```
Message: Resource temporarily unavailable
SocketErrorCode: WouldBlock
```

Reply:
There is something strange. When in the above code 
```diff
server.Blocking = false;
+DumpState(server);
```
```c#
static void DumpState(Socket socket)
{
    BindingFlags flags = BindingFlags.Instance | BindingFlags.NonPublic;
    FieldInfo _willBlock = typeof(Socket).GetField("_willBlock", flags);
    FieldInfo _willBlockInternal = typeof(Socket).GetField("_willBlockInternal", flags);

    object willBlock = _willBlock.GetValue(socket);
    object willBlockInternal = _willBlockInternal.GetValue(socket);

    Console.WriteLine($"Blocking: {socket.Blocking}\nwillBlock: {willBlock}\nwillBlockInternal: {willBlockInternal}\n");
}
```
Then on Windows it's the same behaviour. It runs fine whether `server.Blocking` is true/false.

On Linux it gets flaky with this line. Means on some runs it fails immediately (expected), on some runs it finished without failure or it fails somewhere in the middle.
Seems there's a race-condition anywhere?

Reply:
Interesting.

We shouldn't block this PR on this issue, since it already exists. I'll file a separate issue.

Reply:
That's very strange. I'm not surprised it's flaky on Linux, but I am surprised that whether you call "DumpState" or not determines whether it is flaky. Very odd.

Reply:
Issue here: https://github.com/dotnet/runtime/issues/47287